### PR TITLE
Add `plot_hinton()`

### PR DESCRIPTION
**Pull request by pierreguilmin**
Fixes DYN-44. It wasn't really a priority but I had the code ready, and @abocquet told me he might need this soon. Turned out to take longer that expected. 😅 Anyway, there it is.

I picked sane defaults for the colorbar, square areas and colors for the three following type of input:
- real-valued matrix
- real-valued matrix with only positive values
- complex matrix

@gautierronan I need your help for two things before merging:
- [x] Find/make a nicer cyclic colorbar than the `'twilight'` ugly one of matplotib, we need a cyclic colorbar because for complex numbers the -pi phase should have the same color as the pi phase. Ideally we don't want to have white in this cyclic colormap, bc the background of the plot is white.
- [x] Fix the last commit, there is an issue when trying to render the table of functions in the API apge when I add `plot_hinton` to the list, I guess related to the way you do fancy magic with Jinja templating.

<details>
  <summary>Traceback</summary>

```shell
ERROR   -  Error reading page 'python_api/index.md': No first item, sequence was empty.
Traceback (most recent call last):
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/bin/mkdocs", line 8, in <module>
    sys.exit(cli())
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/mkdocs/__main__.py", line 270, in serve_command
    serve.serve(**kwargs)
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/mkdocs/commands/serve.py", line 86, in serve
    builder(config)
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/mkdocs/commands/serve.py", line 67, in builder
    build(config, live_server=None if is_clean else server, dirty=is_dirty)
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/mkdocs/commands/build.py", line 322, in build
    _populate_page(file.page, config, files, dirty)
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/mkdocs/commands/build.py", line 175, in _populate_page
    page.render(config, files)
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/mkdocs/structure/pages.py", line 271, in render
    self.content = md.convert(self.markdown)
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/markdown/core.py", line 264, in convert
    root = self.parser.parseDocument(self.lines).getroot()
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/markdown/blockparser.py", line 90, in parseDocument
    self.parseChunk(self.root, '\n'.join(lines))
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/markdown/blockparser.py", line 105, in parseChunk
    self.parseBlocks(parent, text.split('\n\n'))
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/markdown/blockparser.py", line 123, in parseBlocks
    if processor.run(parent, blocks) is not False:
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/mkdocstrings/extension.py", line 125, in run
    html, handler, data = self._process_block(identifier, block, heading_level)
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/mkdocstrings/extension.py", line 221, in _process_block
    rendered = handler.render(data, options)
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/mkdocstrings_handlers/python/handler.py", line 322, in render
    return template.render(
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/jinja2/environment.py", line 1301, in render
    self.environment.handle_exception()
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/jinja2/environment.py", line 936, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "docs/templates/python/material/module.html", line 54, in top-level template code
    {% include "table.html" with context %}
  File "docs/templates/python/material/table.html", line 35, in top-level template code
    {% set text = section.value|convert_markdown(heading_level, html_id) %}
  File "/Users/pierreguilmin/.pyenv/versions/dynamiqs/lib/python3.9/site-packages/jinja2/environment.py", line 485, in getattr
    return getattr(obj, attribute)
jinja2.exceptions.UndefinedError: No first item, sequence was empty.
```

</details>

Various remarks:
1. (added to linear) We need to fix the `add_colorbar` function, it does not work well with multiple subplots.
2. (commented on the corresponding linear issue) Once we have the `rand_dm`/`rand_herm` methods we can use them for more meaningfull examples for the plots function.

Here's a rendering of the doc page (I just discovered "Capture full page" in the Arc web broswer command palette 😍):
plot_hinton - dynamiqs (1)

### CODE REVIEW

**Review by abocquet — APPROVED** [review]
Looks good! Could you add a `tickslabel` option? We don't always want to plot the hinton plot in fock basis but sometimes in logical or tensor product basis

**Comment by pierreguilmin** [comment]
> Looks good! Could you add a `tickslabel` option? We don't always want to plot the hinton plot in fock basis but sometimes in logical or tensor product basis

Is it the same labelling for x and y then? Could you give two quick example of the API for `tickslabel` for both case?

**Comment by abocquet** [comment]
Yes they should be the same
```python
dq.hinton(rho, tickslabel=["|00>", "|01>", "|10>", "|11>"])
```
seems good

**Comment by gautierronan** [comment]
@pierreguilmin Cyclic colormap proposal.

```
cmap_colors = [
    (0.0, '#07689D'),
    (0.25, '#AC98AB'),
    (0.5, '#C62525'),
    (0.75, '#5E1A5B'),
    (1.0, '#07689D'),
]
```

Screenshot 2023-11-24 at 12 48 36

**Comment by pierreguilmin** [comment]
Thanks @gautierronan! Random thought: I love the automatic display for colormaps in jupyter notebooks
Screenshot 2023-11-24 at 13 34 40

**Comment by abocquet** [comment]
I find it hard to see the constrasts with your proposition @gautierronan. Maybe we could have complementary colors for phases that are opposite ?

File: dynamiqs/plots/__init__.py [hunk 1]
```diff
@@ -4,12 +4,24 @@
 from .namespace import *
 from .utils import *
 
+# diverging
 cmap_colors = [
     (0.0, '#05527B'),
-    (0.225, '#639DC1'),
+    (0.25, '#639DC1'),
```
Comment (gautierronan) [inline dynamiqs/plots/__init__.py:10]: I had adjusted these values so the map was linearly increasing in lightness

**Comment by gautierronan** [comment]
> I find it hard to see the constrasts with your proposition @gautierronan. Maybe we could have complementary colors for phases that are opposite ?

It is indeeed possible to have complementary for opposite phases, whichever the phases, see the hue colormap in https://stackoverflow.com/a/34557535. I don't find it particularly visually pleasing, so I instead went with the twilight colormap styling: two complementary colors for phase 0 and phase pi (here, red and blue), and a single color in the middle, either dark or light (here, purple).

We can go with the hue colormap if you prefer.

File: dynamiqs/plots/__init__.py (hunk 1 above)
Comment (pierreguilmin) [reply dynamiqs/plots/__init__.py:10]: Hooo, ok sorry, I'll fix it back.

**Comment by gautierronan** [comment]
> Fix the last commit, there is an issue when trying to render the table of functions in the API apge when I add plot_hinton to the list, I guess related to the way you do fancy magic with Jinja templating.

@pierreguilmin That's because the jinja was looking for the first sentence in the docstring, but there wasn't one. I simply added a single-sentence description to your function. We could otherwise write a bit of jinja to check whether there is actually a sentence. But then it doesn't make so much sense to put the function in the API table.

**Review by gautierronan — DISMISSED** [review]

**Comment by pierreguilmin** [comment]
> Yes they should be the same
> 
> ```python
> dq.hinton(rho, tickslabel=["|00>", "|01>", "|10>", "|11>"])
> ```
> 
> seems good

I added automatic ket/bra labelling (if you plot an operator $O$, the column labelled $\ket n$ are the values $O\ket{n}$ and the row labelled $\bra n$ are the values $\bra{n}O$, and one element at column $\ket n$, row $\ket m$ is the value $\braket{m|O|n}$.

Screenshot 2023-11-26 at 19 57 05

**Review by gautierronan — DISMISSED** [review]

**Comment by abocquet** [comment]
> It is indeeed possible to have complementary colors for opposite phases, whichever the phase. This is for instance the hue colormap in https://stackoverflow.com/a/34557535. I don't find it particularly pleasing visually, so I instead went with the twilight colormap styling: two complementary colors for phase 0 and phase pi (here, red and blue), and a single color in the middle (here, purple), either dark for phase pi/2 or light for phase -pi/2.
>We can go with the hue colormap if you prefer.

@gautierronan yes I would strongly prefer hue. I agree it may not be the most beautiful but it by far the most convenient. I've been staring at hinton plots for a few days now an it's nice to able quickly able to reason about colors to find coefficients that have an opposite phase. If you really are sad with this we could propose two cmaps, one handy and one good looking.

**Comment by gautierronan** [comment]
The function already supports a custom `cmap` argument, so it's not a matter of having two supported, but rather of choosing the one we want by default. I'm not arguing for one or the other, simply Pierre asked me for a custom colormap so I made one...

**Comment by pierreguilmin** [comment]
I think the [CMasher package](https://cmasher.readthedocs.io/index.html) has some nice cyclic colormap (we'd need to change the figure background color for the two last ones):
18187fcd-8285-4908-a78d-6f0b5cfa17e6

@abocquet any you like? (copper is my favorite)

**Pull request merged**